### PR TITLE
Refactor Neural Network

### PR DESCRIPTION
**Pull request by ftalbrecht**
This

* puts training to `pymor.algorithms.neural_network`
* adds `StationaryNeuralNetworkModel` which behaves like `StationaryModel` except that `solve` uses the NN, not the operator inversion
* adds `StationaryNeuralNetworkRBReductor` deriving from `StationaryRBReductor`

The latter in particular
* allows to specify a custom RB to reduce onto
* can be used in other algorithms, such as a greedy. in each reduce, the fom is first projected onto the RB, then a NN is trained for the current basis (this requires all arguments to reduce to be provided in init)
* allows for the resulting rom to have the same functionality as an RB rom (output, products, norms)

This allows for a more pyMOR-ic use, as in a comparison of the NN approximation vs. the RB approximation as in

```python
# create some training data
training_mus = parameter_space.sample_uniformly(10)
training_snapshots = fom.solution_space.empty()
for mu in training_mus:
    training_snapshots.append(fom.solve(mu))
training_data = (training_mus, training_snapshots)

RB, _ = pod(training_snapshots, l2_err=5e-6)

validation_mus = parameter_space.sample_randomly(3)
validation_snapshots = fom.solution_space.empty()
for mu in validation_mus:
    validation_snapshots.append(fom.solve(mu))
validation_data = (validation_mus, validation_snapshots)

# generate the ROM
from pymor.reductors.neural_network import StationaryNeuralNetworkRBReductor

reductor = StationaryNeuralNetworkRBReductor(fom, training_data, validation_data, RB, ann_mse=None)
rom = reductor.reduce()

# compare
for mu in parameter_space.sample_randomly(3):
    u_nn = rom.solve(mu)
    u_rb = rom.operator.apply_inverse(rom.rhs.as_range_array(), mu=mu)
    U = fom.solve(mu)
    print(f'NN H^1 error: {fom.h1_0_norm(U - reductor.reconstruct(u_nn))}')
    print(f'RB H^1 error: {fom.h1_0_norm(U - reductor.reconstruct(u_rb))')
```

There are still manay open questions and details to be sorted out, this is just a first draft...

### CODE REVIEW

File: src/pymor/reductors/neural_network.py [hunk 1]
```diff
@@ -16,10 +18,92 @@
...
+
+    class StationaryNeuralNetworkRBReductor(StationaryRBReductor):
+        '''
+        assumes RB to be orthogonal w.r.t. product
+        '''
+
+        def __init__(self, fom, training_data, validation_data,
+                RB=None, product=None, check_orthonormality=None, check_tol=None,
+                hidden_layers='[(N+P)*3, (N+P)*3]', ann_mse='like_basis',
+                max_restarts=10, torch_seed=None, ann_train_params=None):
```
Comment (HenKlei) [inline src/pymor/reductors/neural_network.py:38]: The nice thing about the `NeuralNetworkReductor` was that one could call it with different parameters. For instance, if the loss obtained with a trained network was not sufficient, one could have changed the number of hidden layers in an automatic routine to (hopefully) improve the training results. When having to pass all those information to `__init__`, this is not possible anymore. However, I understand that this is required here, if we do not want to override the `reduce` method of `ProjectionBasedReductor`.

File: src/pymor/reductors/neural_network.py [hunk 2]
```diff
@@ -106,6 +107,70 @@ def data_to_tensor(data):
...
+            ann_train_params.pop('seed', None) # has to be removed, otherwise each training uses the same
+            if torch_seed:
+                torch.manual_seed(torch_seed)
+
+            # determine the numbers of neurons in the hidden layers
+            if isinstance(hidden_layers, str):
+                hidden_layers = eval(hidden_layers, {'P': fom.parameters.dim})
+            else:
+                hidden_layers = self.hidden_layers
+            assert isinstance(hidden_layers, (list, tuple)) and len(hidden_layers) == 2
```
Comment (HenKlei) [inline src/pymor/reductors/neural_network.py:145]: Why `len(hidden_layers) == 2`? Even `len(hidden_layers) == 0` would be totally fine.

File: src/pymor/reductors/neural_network.py [hunk 3]
```diff
@@ -106,6 +107,70 @@ def data_to_tensor(data):
...
+            assert all(isinstance(l, Number) for l in hidden_layers)
+            self.layers = [len(fom.parameters),] + hidden_layers + [fom.output_space.dim,]
+
+            self.__auto_init(locals())
+
+        def reduce(self):
```
Comment (HenKlei) [inline src/pymor/reductors/neural_network.py:151]: Why not in `build_rom` this time, as in the `StationaryNeuralNetworkRBReductor`?

File: src/pymor/algorithms/neural_network.py [hunk 4]
```diff
@@ -0,0 +1,296 @@
...
+
+        # LBFGS-optimizer does not support mini-batching, so the batch size needs to be adjusted
+        if optimizer == optim.LBFGS:
+            batch_size = max(len(training_data), len(validation_data))
+
+        logger.info('Initializing neural network ...')
+
+        # initialize the neural network
+        neural_network = FullyConnectedNN(layers,
+                activation_function=activation_function).double()
```
[the remark this answers is not in the document]
Comment (ftalbrecht) [reply src/pymor/algorithms/neural_network.py:43]: True. Perhaps we make `neural_network=None` optional and create it if `layers` and `activation_function` is given?

File: src/pymor/reductors/neural_network.py (hunk 1 above)
Comment (ftalbrecht) [reply src/pymor/reductors/neural_network.py:39]: I am for instance thinking about a greedy algorithm which calls `reduce()` on the same reductor in each iteration after extending the basis. There, we do nat want the greedy to need to know about the kind of reductor. However, we now have the surrogates as input for the greedy which are there for exactly that reason. So, its not that clear. We should probably we involve the other pyMOR devs in the discussion at some point... :)

File: src/pymor/reductors/neural_network.py (hunk 2 above)
Comment (ftalbrecht) [reply src/pymor/reductors/neural_network.py:145]: I have no idea whats correct there, so I will just follow your sugestions!

File: src/pymor/algorithms/neural_network.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (HenKlei) [reply src/pymor/algorithms/neural_network.py:43]: Sounds good. So as default, we create `neural_network` as `FullyConnectedNN`. Later, one could also think about a `ConvolutionalNN` that is passed to the `train_ann` algorithm directly.

Alternatively, one could choose `NeuralNetworkType=FullyConnectedNN` as parameter for `train_ann` and instantiate the neural network in the training algorithm. Then, I would replace `layers` and `activation_function` with a dictionary `neural_network_parameters` and call something like
```
neural_network = NeuralNetworkType(*neural_network_parameters).double()
```
This would perhaps be the most flexible way if we do not want to instantiate the neural network beforehand. For convolutional neural networks we would call `train_ann(..., NeuralNetworkType=ConvolutionalNN, neural_network_parameters={'layers': [...], 'activation_function': ..., some other parameters specific for convolutional neural networks})`.

File: src/pymor/reductors/neural_network.py (hunk 1 above)
Comment (HenKlei) [reply src/pymor/reductors/neural_network.py:39]: And the neural network architecture would be adjusted whenever calling `reduce()` since the required output dimension of the neural network might have changed, right?

File: src/pymor/reductors/neural_network.py (hunk 3 above)
Comment (ftalbrecht) [reply src/pymor/reductors/neural_network.py:151]: I Since we do not derive from the `StationaryRBReductor` there is only `reduce`, not `build_rom`. But even then, where to do all the work is debatable. I think that `reduce` (respectively `build_rom`) should contain all non-trivial work, while sanity checks and other configs are best carried out in init. Or did I not understand your question correctly?

File: src/pymor/reductors/neural_network.py (hunk 2 above)
Comment (HenKlei) [reply src/pymor/reductors/neural_network.py:145]: The list of hidden layers is extended by the number of inputs and the number of outputs anyway, since these values are fixed and can not be adjusted by the user. It is not necessary to deploy any hidden layers at all. Without a hidden layer, however, the function implemented by the neural network is a linear function, since no (non-linear) activation function is applied in front of the last layer in our implementation.

File: src/pymor/reductors/neural_network.py (hunk 1 above)
Comment (ftalbrecht) [reply src/pymor/reductors/neural_network.py:39]: Exactly, since the RB and its size changed. But this could also mean that a different number of restarts is appropriate for different reduce calls. So its really not fully clear to me where to put things...

File: src/pymor/reductors/neural_network.py (hunk 3 above)
Comment (HenKlei) [reply src/pymor/reductors/neural_network.py:151]: Okay, I see. I was not aware that this reductor does not derive from any other reductor. Then this is fine.

File: src/pymor/reductors/neural_network.py [hunk 5]
```diff
@@ -106,6 +107,70 @@ def data_to_tensor(data):
             return StationaryNeuralNetworkModel(self.neural_network, error_estimator=None, **projected_operators)
 
 
+    class StationaryNeuralNetworkOutputReductor(BasicObject):
+
+        # TODO: allow for transformation/back_transformation of outputs (and inputs?)!
+        def __init__(self, fom, training_data, validation_data,
+                     hidden_layers='[P*3, P*3]', ann_mse=None, max_restarts=10, torch_seed=None, ann_train_params=None):
```
Comment (HenKlei) [inline src/pymor/reductors/neural_network.py:114]: Maybe `hidden_layers` could (by default) also depend on the number of outputs, as it was the case in the `StationaryNeuralNetworkReductor`? For instance something like
```suggestion
                     hidden_layers='[(O+P)*3, (O+P)*3]', ann_mse=None, max_restarts=10, torch_seed=None, ann_train_params=None):
```

File: src/pymor/reductors/neural_network.py [hunk 6]
```diff
@@ -106,6 +107,70 @@ def data_to_tensor(data):
...
+            assert isinstance(ann_train_params, dict)
+            if not torch_seed and 'seed' in ann_train_params:
+                torch_seed = ann_train_params['seed']
+            ann_train_params.pop('seed', None) # has to be removed, otherwise each training uses the same
+            if torch_seed:
+                torch.manual_seed(torch_seed)
+
+            # determine the numbers of neurons in the hidden layers
+            if isinstance(hidden_layers, str):
+                hidden_layers = eval(hidden_layers, {'P': fom.parameters.dim})
```
Comment (HenKlei) [inline src/pymor/reductors/neural_network.py:142]: And then we adjust the `eval` call accordingly:
```suggestion
                hidden_layers = eval(hidden_layers, {'P': fom.parameters.dim, 'O': fom.output_space.dim})
```

File: src/pymor/reductors/neural_network.py (hunk 1 above)
Comment (HenKlei) [reply src/pymor/reductors/neural_network.py:39]: Yes, that makes everything a bit difficult. For instance, the architecture of the neural network should be adjusted as well when increasing the number of outputs in that case (at least in general). Hence, the original neural network architecture and training procedure might be inappropriate when changing the training data. Some time ago, I discussed with @sdrave whether it is possible to implement some sort of general "fitting" routine that adjusts neural network architecture and training parameters automatically to obtain a sufficient model. But I think this is quite difficult... And such a routine might increase the training time dramatically.

File: src/pymor/reductors/neural_network.py (hunk 3 above)
Comment (HenKlei) [reply src/pymor/reductors/neural_network.py:151]: Okay, I see. I was not aware that this new reductor does not derive from any other reductor. Then this is totally fine.

File: src/pymor/reductors/neural_network.py (hunk 6 above)
Comment (ftalbrecht) [reply src/pymor/reductors/neural_network.py:142]: Actually, I have the same change locally by now! :)

However, training for the output seems not to work too well in my example, even when using 4 hidden layers with 100 neurons each (rescaling output did not help either), but this is a different story.

File: src/pymor/reductors/neural_network.py (hunk 2 above)
Comment (ftalbrecht) [reply src/pymor/reductors/neural_network.py]: Finally understood this and updated the checks accordingly, thx.

File: src/pymor/reductors/neural_network.py (hunk 6 above)
Comment (ftalbrecht) [reply src/pymor/reductors/neural_network.py]: Included suggestions.

**Comment by ftalbrecht** [comment]
Please have a look at the update demo. I included an appropriate product in the POD depending on the discretizer (euklidean is not the best choice for P1-CG FEM, I think). I also added a balance of the desired MSE of the NN and the L2-error of the POD, since (if I understand this correctly and the NN is actually trained to fulfill a mean squared error, not an l2-error), we have
```
desired_l2_err = np.sqrt(len(training_snapshots)*desired_MSE)
```
Does this make sense to you?

**Comment by HenKlei** [comment]
> Does this make sense to you?

Yes, after a first look I would say that you are right. The other changes also look quite reasonable.

File: src/pymor/algorithms/neural_network.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (HenKlei) [reply src/pymor/algorithms/neural_network.py:44]: @ftalbrecht What do you think about this?

**Comment by HenKlei** [comment]
What about `validation_ratio`? I think that was quite useful. One just had to provide a single set of parameter/solution pairs and the set was split into training and validation samples according to `validation_ratio`.

**Comment by ftalbrecht** [comment]
> What about `validation_ratio`? I think that was quite useful. One just had to provide a single set of parameter/solution pairs and the set was split into training and validation samples according to `validation_ratio`.

I guess it really depends on the circumstances. I often have a `training_set` lying around that I use as input for different algorithms, where I want them all to train w.r.t. the same set. In that case, I explicitly want to provide separate sets. When used standalone, it seems more convenient to provide only one set and prescribe `validation_ratio`. So if both can be supported, go ahead.

**Comment by HenKlei** [comment]
> I guess it really depends on the circumstances. I often have a `training_set` lying around that I use as input for different algorithms, where I want them all to train w.r.t. the same set. In that case, I explicitly want to provide separate sets. When used standalone, it seems more convenient to provide only one set and prescribe `validation_ratio`. So if both can be supported, go ahead.

I did not plan to remove the possibility of providing separate training and validation sets. I was just wondering if there is a reason to remove `validation_ratio` and thereby the possibility to use a single sample set. We can support both ways, I think, so I will add the `validation_ratio` again.
